### PR TITLE
Added add child page

### PR DESCRIPTION
**Pull request by DanielBrierton**
Also improves usage of firstName/lastName and datepickers for DoB.

### CODE REVIEW

File: web/public/js/directives/cd-add-child/template.dust [hunk 1]
```diff
@@ -0,0 +1,70 @@
...
+  </div>
+  <div class="cd-add-child__o13-section" ng-show="$ctrl.profile.dob < $ctrl.thirteenYearsAgo">
+    <div class="col-sm-10 col-sm-offset-2 cd-add-child__info-box">
+      <p>{@i18n key="As your child is over 13 years of age, they can have their own CoderDojo account."/}</p>
+      <p>{@i18n key="This is optional, but will allow your child to log in by themselves to access features such as badging, booking and the youth forums by themselves. Enter your child's email address, and give them a password if you wish to do this."/}</p>
+    </div>
+    <div class="form-group">
+      <label for="cd-add-child__email" class="col-sm-2 control-label">Email</label>
+      <div class="col-sm-10">
+        <input type="email" class="form-control" name="cd-add-child__email" placeholder="Email" ng-model="$ctrl.profile.email" validator="$ctrl.emailValidator()" invalid-message="'{@i18n key="You must provide an email address when providing a password."/}'">
```
Comment (Wardormeur) [inline web/public/js/directives/cd-add-child/template.dust]: You can use this if the validator is supposed to be reusable https://docs.angularjs.org/api/ng/type/ngModel.NgModelController#$validators

File: web/public/js/services/user-utils.js [hunk 2]
```diff
@@ -13,6 +13,17 @@ angular.module('cpZenPlatform').factory('userUtils',
     return moment.utc().diff(moment(birthDate), 'years');
   };
 
+  userUtils.getBaseUserTypeByAge = function (birthDate) {
```
Comment (Wardormeur) [inline web/public/js/services/user-utils.js]: yes please :)

File: web/public/js/directives/user/cd-profile/children-list/style.less [hunk 3]
```diff
@@ -0,0 +1,5 @@
+.cd-children-list {
+  &__popover {
+    min-width: 290px;
```
Comment (Wardormeur) [inline web/public/js/directives/user/cd-profile/children-list/style.less]: responsive?

File: web/public/js/services/cd-api.js [hunk 4]
```diff
@@ -48,7 +48,12 @@ angular.module('cpZenPlatform').service('cdApi', ['$http', function($http) {
     if (resolve) {
       fn()
       .then(function (response) {
-        return resolve(_.has(response, 'data') && _.has(response, 'headers') && _.has(response, 'config') ? response.data : response);
+        if (_.has(response, 'data') && _.has(response, 'headers') && _.has(response, 'config')) {
+          return resolve(response.data);
+        } else {
+          return resolve(response);
+        }
+        // return resolve(_.has(response, 'data') && _.has(response, 'headers') && _.has(response, 'config') ? response.data : response);
```
Comment (Wardormeur) [inline web/public/js/services/cd-api.js]: remove comment :)

File: web/public/js/directives/user/cd-profile/children-list/index.js [hunk 5]
```diff
@@ -0,0 +1,30 @@
+
+
+;(function() {
+  'use strict';
+
+  function listOfChildren(){
+    return {
+      restrict: 'E',
+      templateUrl: '/directives/tpl/user/cd-profile/children-list', // '/profiles/template/children-list',
```
Comment (Wardormeur) [inline web/public/js/directives/user/cd-profile/children-list/index.js]: i guess it's the right url, but remove comment? :dancing_men:

File: web/public/js/services/cd-users-service.js [hunk 6]
```diff
@@ -42,6 +42,13 @@ function cdUsersService(cdApi, $q){
         cdApi.post('profiles/youth/create', {profile: profile}, win, fail || topfail);
       }
     },
+    saveYouthProfilePromise: function(profile) {
```
Comment (Wardormeur) [inline web/public/js/services/cd-users-service.js]: Not sure if I don't prefer a separation between both, let's talk abt it tomorrow

File: web/public/templates/profiles/general-info.dust [hunk 7]
```diff
@@ -12,8 +12,11 @@
 
         <div class="form-group" ng-enter="noop()">
           <label for="fullname" class="col-sm-4 control-label">{@i18n key="Name"/}* </label>
-          <div class="col-sm-8">
-            <input name="fullname" type="text" required required-message="'Name is required'" class="form-control" id="fullname" placeholder="{@i18n key="Name"/}" ng-model="profile.name">
+          <div class="col-sm-4">
+            <input name="firstName" type="text" required required-message="'Name is required'" class="form-control" id="firstName" placeholder="{@i18n key="Name"/}" ng-model="profile.firstName">
```
Comment (Wardormeur) [inline web/public/templates/profiles/general-info.dust]: i18N on required-message

File: web/public/templates/profiles/general-info.dust [hunk 8]
```diff
@@ -12,8 +12,11 @@
...
+          </div>
+          <div class="col-sm-4">
+            <input name="lastName" type="text" required required-message="'Surname is required'" class="form-control" id="lastName" placeholder="{@i18n key="Surname"/}" ng-model="profile.lastName">
```
Comment (Wardormeur) [inline web/public/templates/profiles/general-info.dust]: same here, i18n :)

File: web/public/js/directives/cd-child-card/template.dust [hunk 9]
```diff
@@ -60,13 +60,13 @@
...
                 <h4>{{ 'MEMBER_OF' | translateFromKey }}</h4>
-                <p>{{ 'REMOVE_CHILD_FROM_DOJO_TEXT' | translateFromKey:{name: $ctrl.child.name} }}</p>
+                <p>{{ 'REMOVE_CHILD_FROM_DOJO_TEXT' | translateFromKey:{name: $ctrl.child.firstName || $ctrl.child.name} }}</p>
               </div>
               <cd-picture-grid ng-show="$ctrl.child.dojoListings.length > 0" items="$ctrl.child.dojoListings" on-selection="$ctrl.manageDojos ? $ctrl.removeFromDojo : $ctrl.noop" item-icon="minus" item-icon-color="red" show-item-icons="$ctrl.manageDojos" align="left" size="medium"></cd-picture-grid>
             </div>
             <div class="col-xs-12 col-md-6" ng-show="$ctrl.manageDojos">
               <h4>{{ 'AVAILABLE_DOJOS' | translateFromKey }}</h4>
-              <p ng-bind-html="'ADD_CHILD_TO_DOJO_TEXT' | translateFromKey:{name: $ctrl.child.name, findADojoUrl: '/'}:'escapeParameters'"></p>
+              <p ng-bind-html="'ADD_CHILD_TO_DOJO_TEXT' | translateFromKey:{name: $ctrl.child.firstName || $ctrl.child.name, findADojoUrl: '/'}:'escapeParameters'"></p>
```
Comment (Wardormeur) [inline web/public/js/directives/cd-child-card/template.dust]: possible xss due to binding of a username into a html-binded area?

File: web/locale/en_US/messages.po [hunk 10]
```diff
@@ -3250,3 +3250,33 @@ msgstr "%1s at %2s on %3s"
 msgid "%1s in %2s"
 msgctxt "Metadata displayed when sharing a dojo page, %1s being the dojo name, %2s being the country"
 msgstr "%1s in %2s"
+
+msgid "First Name"
```
Comment (Wardormeur) [inline web/locale/en_US/messages.po]: Update messages.po to avoid overwrite ? :)

File: web/public/js/directives/cd-child-card/template.dust (hunk 9 above)
Comment (DanielBrierton) [inline web/public/js/directives/cd-child-card/template.dust]: the `:'escapeParameters'` at the end... escapes the parameters... 😝  So HTML from firstName || name will be escaped.

File: web/public/js/services/cd-users-service.js (hunk 6 above)
Comment (DanielBrierton) [inline web/public/js/services/cd-users-service.js]: Fair point. Just have to add returns to the original message and cd-api will handle it for us.

**Pull request merged**